### PR TITLE
esp32s3/spi-dma: Fix spi dma transfer.

### PR DESCRIPTION
**Pull request by w2016561536**
## Summary
Fix esp32s3 spi dma transfer.

## Impact
Fix esp32s3 spi dma transfer only transmit first byte and receive empty problem.

## Testing
Base config: esp32s3-devkit:spi  
Changes: 
```
CONFIG_DEBUG_FEATURES=y
CONFIG_DEBUG_SPI=y
CONFIG_DEBUG_SPI_ERROR=y
CONFIG_DEBUG_SPI_INFO=y
CONFIG_DEBUG_SPI_WARN=y
CONFIG_ESP32S3_SPI2_CLKPIN=2
CONFIG_ESP32S3_SPI2_CSPIN=1
CONFIG_ESP32S3_SPI2_MISOPIN=41
CONFIG_ESP32S3_SPI2_MOSIPIN=42
CONFIG_ESP32S3_SPI3_CLKPIN=13
CONFIG_ESP32S3_SPI3_CSPIN=11
CONFIG_ESP32S3_SPI3_MISOPIN=12
CONFIG_ESP32S3_SPI3_MOSIPIN=14
CONFIG_ESP32S3_SPI_DMA=y
CONFIG_ESP32S3_SPI_DMATHRESHOLD=4
```

defconfig after edit : 
``` 
#
# This file is autogenerated: PLEASE DO NOT EDIT IT.
#
# You can use "make menuconfig" to make any modifications to the installed .config file.
# You can then do "make savedefconfig" to generate a new defconfig file that includes your
# modifications.
#
# CONFIG_ARCH_LEDS is not set
# CONFIG_NSH_ARGCAT is not set
# CONFIG_NSH_CMDOPT_HEXDUMP is not set
CONFIG_ARCH="xtensa"
CONFIG_ARCH_BOARD="esp32s3-devkit"
CONFIG_ARCH_BOARD_COMMON=y
CONFIG_ARCH_BOARD_ESP32S3_DEVKIT=y
CONFIG_ARCH_CHIP="esp32s3"
CONFIG_ARCH_CHIP_ESP32S3=y
CONFIG_ARCH_CHIP_ESP32S3WROOM1=y
CONFIG_ARCH_INTERRUPTSTACK=2048
CONFIG_ARCH_STACKDUMP=y
CONFIG_ARCH_XTENSA=y
CONFIG_BOARD_LOOPSPERMSEC=16717
CONFIG_BUILTIN=y
CONFIG_DEBUG_FEATURES=y
CONFIG_DEBUG_SPI=y
CONFIG_DEBUG_SPI_ERROR=y
CONFIG_DEBUG_SPI_INFO=y
CONFIG_DEBUG_SPI_WARN=y
CONFIG_ESP32S3_SPI2=y
CONFIG_ESP32S3_SPI2_CLKPIN=2
CONFIG_ESP32S3_SPI2_CSPIN=1
CONFIG_ESP32S3_SPI2_MISOPIN=41
CONFIG_ESP32S3_SPI2_MOSIPIN=42
CONFIG_ESP32S3_SPI3=y
CONFIG_ESP32S3_SPI3_CLKPIN=13
CONFIG_ESP32S3_SPI3_CSPIN=11
CONFIG_ESP32S3_SPI3_MISOPIN=12
CONFIG_ESP32S3_SPI3_MOSIPIN=14
CONFIG_ESP32S3_SPI_DMA=y
CONFIG_ESP32S3_SPI_DMATHRESHOLD=4
CONFIG_ESP32S3_UART0=y
CONFIG_FS_PROCFS=y
CONFIG_HAVE_CXX=y
CONFIG_HAVE_CXXINITIALIZE=y
CONFIG_IDLETHREAD_STACKSIZE=3072
CONFIG_INIT_ENTRYPOINT="nsh_main"
CONFIG_INTELHEX_BINARY=y
CONFIG_NSH_ARCHINIT=y
CONFIG_NSH_BUILTIN_APPS=y
CONFIG_NSH_FILEIOSIZE=512
CONFIG_NSH_LINELEN=64
CONFIG_NSH_READLINE=y
CONFIG_PREALLOC_TIMERS=4
CONFIG_RAM_SIZE=114688
CONFIG_RAM_START=0x20000000
CONFIG_RR_INTERVAL=200
CONFIG_SCHED_WAITPID=y
CONFIG_SPITOOL_MINBUS=2
CONFIG_START_DAY=6
CONFIG_START_MONTH=12
CONFIG_START_YEAR=2011
CONFIG_SYSLOG_BUFFER=y
CONFIG_SYSTEM_NSH=y
CONFIG_SYSTEM_SPITOOL=y
CONFIG_UART0_SERIAL_CONSOLE=y
``` 
### Test in cli: 
1. On spi 2, use poll method to read icm-20602 chip id.
icm20602-poll-cli
On hardware:
icm20602-poll

2. On spi 2, use dma method to read icm-20602 chip id.
icm20602-dma-cli
On hardware:
icm20602-dma

3. On spi 3, use poll to read FM25V05 chip id:
fm25-poll-cli
On hardware:
fm25-poll

4. On spi 3, use dma to read FM25V05 chip id:
fm25-dma-cli
On hardware: 
fm25-dma

### CODE REVIEW

File: arch/xtensa/src/esp32s3/esp32s3_spi.c [hunk 1]
```diff
@@ -979,6 +988,14 @@ static void esp32s3_spi_dma_exchange(struct esp32s3_spi_priv_s *priv,
 static uint32_t esp32s3_spi_poll_send(struct esp32s3_spi_priv_s *priv,
                                       uint32_t wd)
 {
+#ifdef CONFIG_ESP32S3_SPI_DMA
+  esp32s3_spi_clr_regbits(SPI_DMA_CONF_REG(priv->config->id),
+                              SPI_DMA_TX_ENA_M);
```
Comment (xiaoxiang781216) [inline arch/xtensa/src/esp32s3/esp32s3_spi.c:993]:
```suggestion
                          SPI_DMA_TX_ENA_M);
```

File: arch/xtensa/src/esp32s3/esp32s3_spi.c [hunk 2]
```diff
@@ -979,6 +988,14 @@ static void esp32s3_spi_dma_exchange(struct esp32s3_spi_priv_s *priv,
...
+
+  esp32s3_spi_clr_regbits(SPI_DMA_CONF_REG(priv->config->id),
+                              SPI_DMA_RX_ENA_M);
```
Comment (xiaoxiang781216) [inline arch/xtensa/src/esp32s3/esp32s3_spi.c:996]:
```suggestion
                          SPI_DMA_RX_ENA_M);
```

File: arch/xtensa/src/esp32s3/esp32s3_spi.c [hunk 3]
```diff
@@ -1059,6 +1076,14 @@ static void esp32s3_spi_poll_exchange(struct esp32s3_spi_priv_s *priv,
                                       void *rxbuffer,
                                       size_t nwords)
 {
+#ifdef CONFIG_ESP32S3_SPI_DMA
+  esp32s3_spi_clr_regbits(SPI_DMA_CONF_REG(priv->config->id),
+                              SPI_DMA_TX_ENA_M);
```
Comment (xiaoxiang781216) [inline arch/xtensa/src/esp32s3/esp32s3_spi.c:1081]: ditto

File: arch/xtensa/src/esp32s3/esp32s3_spi.c [hunk 4]
```diff
@@ -1059,6 +1076,14 @@ static void esp32s3_spi_poll_exchange(struct esp32s3_spi_priv_s *priv,
...
+
+  esp32s3_spi_clr_regbits(SPI_DMA_CONF_REG(priv->config->id),
+                              SPI_DMA_RX_ENA_M);
```
Comment (xiaoxiang781216) [inline arch/xtensa/src/esp32s3/esp32s3_spi.c:1084]: ditto

**Review by xiaoxiang781216 — APPROVED** [review]

**Comment by tmedicci** [comment]
I'll double-check it tomorrow. Please don't merge it yet.

**Review by jerpelea — APPROVED** [review]
can you please add a commit message

File: arch/xtensa/src/esp32s3/esp32s3_dma.h [hunk 5]
```diff
@@ -64,7 +64,7 @@ extern "C"
 
 /* DMA channel number */
 
-#define ESP32S3_DMA_CHAN_MAX          (3)
```
Comment (tmedicci) [inline arch/xtensa/src/esp32s3/esp32s3_dma.h:67]: @acassis , do you recall why didn't you assign all 5 GDMA channels to S3? I looked for uses without using the `esp32s3_dma_request` "API" and I couldn't fin anything using GDMA channels directly.

File: arch/xtensa/src/esp32s3/esp32s3_spi.c [hunk 6]
```diff
@@ -1312,22 +1337,23 @@ void esp32s3_spi_dma_init(struct spi_dev_s *dev)
...
   struct esp32s3_spi_priv_s *priv = (struct esp32s3_spi_priv_s *)dev;
 
-  /* Enable GDMA clock for the SPI peripheral */
-
-  modifyreg32(SYSTEM_PERIP_CLK_EN0_REG, 0, priv->config->dma_clk_bit);
-
-  /* Reset GDMA for the SPI peripheral */
-
-  modifyreg32(SYSTEM_PERIP_RST_EN0_REG, priv->config->dma_rst_bit, 0);
-
```
Comment (tmedicci) [inline arch/xtensa/src/esp32s3/esp32s3_spi.c:1322]: Why did you remove device initialization? I recommend reverting it...

Comment (w2016561536) [reply arch/xtensa/src/esp32s3/esp32s3_spi.c:1322]: In technical reference, There aren't these two registers. So i delete it.

Comment (tmedicci) [reply arch/xtensa/src/esp32s3/esp32s3_spi.c:1322]: It's the overall peripheral initialization. You can keep it...

**Comment by tmedicci** [comment]
@w2016561536 , could you please:

1) Fill the description, including the testing method;
2) Apply the following patch to fix whether to use or not the DMA:
```
diff --git a/arch/xtensa/src/esp32s3/esp32s3_spi.c b/arch/xtensa/src/esp32s3/esp32s3_spi.c
index d700fb7c33..af8681b8cd 100644
--- a/arch/xtensa/src/esp32s3/esp32s3_spi.c
+++ b/arch/xtensa/src/esp32s3/esp32s3_spi.c
@@ -1223,7 +1223,7 @@ static void esp32s3_spi_exchange(struct spi_dev_s *dev,
 #ifdef CONFIG_ESP32S3_SPI_DMA
   size_t thld = CONFIG_ESP32S3_SPI_DMATHRESHOLD;
 
-  if (nwords > thld)
+  if ((nwords * priv->nbits) / 8 > thld)
     {
       esp32s3_spi_dma_exchange(priv, txbuffer, rxbuffer, nwords);
     }
```

**Comment by w2016561536** [comment]
I'm working on test now. A detailed description will be added to first conversation soon.

**Comment by w2016561536** [comment]
@tmedicci comment has been updated, please review it.

File: arch/xtensa/src/esp32s3/esp32s3_spi.c (hunk 6 above)
Comment (tmedicci) [reply arch/xtensa/src/esp32s3/esp32s3_spi.c:1322]: Please, revert your changes and keep the overall device initialization.

Comment (w2016561536) [reply arch/xtensa/src/esp32s3/esp32s3_spi.c:1322]: [https://www.espressif.com.cn/sites/default/files/documentation/esp32-s3_technical_reference_manual_en.pdf#page=810](url)
But, in SYSTEM_PERIP_CLK_EN0_REG, bit 22 and bit 27 are reserved. In SYSTEM_PERIP_CLK_EN1_REG, bit 22 and bit 27 are reserved either. In function ` esp32s3_dma_init() ` , dma clock will be enabled and reset status will be removed.

Comment (w2016561536) [reply arch/xtensa/src/esp32s3/esp32s3_spi.c]: > Please, revert your changes and keep the overall device initialization.

OK

**Review by tmedicci — APPROVED** [review]
LGTM. Congrats!

**Comment by w2016561536** [comment]
Hello? Is anyone could help me to merge this pr?
